src/libponyc/expr/operator.c

```diff
@@ -438,8 +438,13 @@ bool expr_assign(pass_opt_t* opt, ast_t* ast)
 
   if(!ok_safe)
   {
-    if(ast_id(left) == TK_FVARREF && ast_child(left) != NULL &&
-      ast_id(ast_child(left)) == TK_THIS)
+    ast_t *left_lastchild = left;
+    while(ast_child(left_lastchild) != NULL)
+    {
+      left_lastchild = ast_child(left_lastchild);
+    }
+    if(ast_id(left) == TK_FVARREF &&
```

Comment:
This may also need to include ast_id(left) being one of:
TK_LETREF and TK_EMBEDREF

Reply:
These should correspond to the types of fields `let` for fields that can't be reassigned (but might be mutated), and `embed` which you may not have seen yet, which stores the data for the class inline.